### PR TITLE
Fix downcast for bounded ints

### LINKED ISSUE #1456: Review downcast libfunc for bounded ints
Review the `downcast` libfunc, to ensure that bounded int values are taken into account

- What types can the libfunc arguments have?
- Is the offset of all bounded int values taken into account?
- Is the value of the return bounded int (if any) computed correctly, according to its offset?

We should:

- Document the signature the signature of the libfunc, including input and output values.
- Document the expected behavior of the libfunc.
- Add unit/integration tests for edge cases.

### PR DESCRIPTION
**Pull request by FrancoGiachetta**
# Fix downcast for bounded ints

This PR fixes a bug on the trivial case of a downcast. This used to be an error before, which was actually wrong since it's not the case for the cairo compiler which returns the same value with the same type.
Additionally, I did two more things:
1.  I Added more tests with different cases of downcast.
2. Simplified the libfunc's implementation. Removed some checks such if both the lower and the upper bound of the source type are included in the target type, which imposible since that would only be the trivial case due to the essence of the downcast operation. 

Closes #1456

## Introduces Breaking Changes?

No.

- [ ] Created PR in [sequencer](https://github.com/lambdaclass/sequencer)
- [ ] Created PR in [starknet-replay](https://github.com/lambdaclass/starknet-replay)   
- [ ] Updated the `starknet-blocks.yml` workflow to use these PRs.

These PRs should be merged after this one right away, in that order.

## Checklist

- [ ] Linked to Github Issue.
- [ ] Unit tests added.
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: src/libfuncs/cast.rs [hunk 1]
```diff
@@ -102,31 +112,16 @@ pub fn build_downcast<'ctx, 'this>(
...
-        } else {
-            entry.extui(
-                src_value,
-                IntegerType::new(context, compute_width).into(),
-                location,
-            )?
-        }
-    } else {
-        src_value
-    };
```
[the remark this answers is not in the document]
Comment (FrancoGiachetta) [reply src/libfuncs/cast.rs]: Well, I came up with a case which makes this actually true: 
```
BoundedInt<0,30> -> BoundedInt<5,40>
```
This is a valid downcast in cairo, so the previous comment is not true.

File: src/libfuncs/cast.rs [hunk 2]
```diff
@@ -46,7 +46,16 @@ pub fn build<'ctx, 'this>(
     }
 }
 
-/// Generate MLIR operations for the `downcast` libfunc.
+/// Generate MLIR operations for the `downcast` libfunc. which converts from a
+/// source type `T` to a target type `U`, where `U` is included in `T`. This
+/// means that the operation can fail.
```
[the remark this answers is not in the document]
Comment (FrancoGiachetta) [reply src/libfuncs/cast.rs]: I would say _U might not fully include T_ since we can perform a downcast from T to T. I'll change it.

File: src/libfuncs/cast.rs [hunk 3]
```diff
@@ -102,30 +112,15 @@ pub fn build_downcast<'ctx, 'this>(
         dst_range.zero_based_bit_width()
     };
 
-    let compute_width = src_range
-        .zero_based_bit_width()
-        .max(dst_range.zero_based_bit_width());
+    let compute_width = src_width.max(dst_width);
```
[the remark this answers is not in the document]
Comment (FrancoGiachetta) [reply src/libfuncs/cast.rs]: Yes, I think I misunderstood the difference with the `zero_based_bit_width`. I'll change it since it makes more sense.

File: src/libfuncs/cast.rs [hunk 4]
```diff
@@ -466,6 +488,79 @@ mod test {
...
+                Option<i32>,
+                Option<i64>,
+            ) {
+                (
+                    downcast(v8),
+                    downcast(v16),
+                    downcast(v32),
+                    downcast(v64),
+                )
+            }
```
Comment (JulianGCalderon) [inline src/libfuncs/cast.rs]: To be consistent, would it be worth changing this test (and the existing test) to be similar to the bounded int tests?

- A single program with a single function for each conversion.
- Using test_case to quickly test multiple cases.

File: src/libfuncs/cast.rs [hunk 5]
```diff
@@ -466,6 +488,79 @@ mod test {
...
+
+            fn run_test_1(val: felt252) -> Option<BoundedInt<0,30>> {
+                let bounded: BoundedInt<0,30> = val.try_into().unwrap();
+                downcast(bounded)
+            }
+
+            fn run_test_2(val: felt252) -> Option<BoundedInt<-31,30>> {
+                let bounded: BoundedInt<-31,30> = val.try_into().unwrap();
+                downcast(bounded)
+            }
```
[the remark this answers is not in the document]
Comment (FrancoGiachetta) [reply src/libfuncs/cast.rs]: Done

File: src/libfuncs/cast.rs [hunk 6]
```diff
@@ -509,25 +581,60 @@ mod test {
                     jit_enum!(1, jit_struct!()),
                     jit_enum!(1, jit_struct!()),
                     jit_enum!(1, jit_struct!()),
-                    jit_enum!(1, jit_struct!()),
```
Comment (JulianGCalderon) [inline src/libfuncs/cast.rs]: Just to confirm, this is because now we don't fail when downcasting to the same type, right?

**Review by JulianGCalderon — APPROVED** [review]
Looks good!

File: src/libfuncs/cast.rs (hunk 6 above)
Comment (FrancoGiachetta) [reply src/libfuncs/cast.rs]: Yes, I've checked it with `cairo-run` this was the actual behaviour, and it returns the same value of the same type.

File: src/libfuncs/cast.rs [hunk 7]
```diff
@@ -108,6 +118,7 @@ pub fn build_downcast<'ctx, 'this>(
 
     let is_signed = src_range.lower.sign() == Sign::Minus;
 
+    // It can happen that the target type is wider then the source type. Extend the value representation.
```
Comment (gabrielbosio) [inline src/libfuncs/cast.rs]:
```suggestion
    // If the target type is wider than the source type, extend the value representation width.
```

**Review by JulianGCalderon — DISMISSED** [review]
LGTM!

File: src/libfuncs/cast.rs [hunk 8]
```diff
@@ -91,6 +89,29 @@ pub fn build_downcast<'ctx, 'this>(
         src_ty.integer_range(registry)?
     };
 
+    // This is the trivial case, so we just return the value.
```
[the remark this answers is not in the document]
Comment (DiegoCivi) [reply src/libfuncs/cast.rs]: I think its okay. That "if" catches the case where a downcast is done from type `A` to type `A`. For example from `BoundedInt<0, 30>` to `BoundedInt<0, 30>`

[the remark this answers is not in the document]
Comment (FrancoGiachetta) [reply src/libfuncs/cast.rs]: No, it is just to clarify that's the trivial case, but it is not outdated.

**Review by DiegoCivi — DISMISSED** [review]

**Review by JulianGCalderon — APPROVED** [review]
LGTM!

File: src/libfuncs/cast.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (gabrielbosio) [reply src/libfuncs/cast.rs]: That would be fine :)

**Review by gabrielbosio — APPROVED** [review]

File: src/libfuncs/cast.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (JulianGCalderon) [reply src/libfuncs/cast.rs]: Done in 147bed23dd4ef19cd2863570aa072807011ad22f.

**Review by DiegoCivi — APPROVED** [review]
